release-23.1: kvserver: eagerly extend expiration leases in Raft scheduler

erikgrinaker opened the pull request:
Backport:
  * 1/1 commits from "kvserver: register lease renewals on splits/merges" (#100392)
  * 4/4 commits from "kvserver: eagerly extend expiration leases in Raft scheduler" (#100430)
  * 2/2 commits from "kvserver: deflake `TestLeaseRenewer`" (#101012)

Please see individual PRs for details.

/cc @cockroachdb/release

Release justification: fixes rangefeed resolved timestamps with expiration leases, gated behind default-off cluster setting.

erikgrinaker committed:
kvserver: don't quiesce ranges with expiration-based leases

This patch prevents quiescence with expiration-based leases, since we'll
have to propose a lease extension shortly anyway. Out of caution, we
only do this when `kv.expiration_leases_only.enabled` is `true`, with an
eye towards 23.1 backports. For 23.2, we will always do this.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: also acquire expiration leases in replicate queue

Previously, the replicate queue would only reacquire epoch leases, and
allow expiration leases to expire. It now reacquires all leases, since
they are eagerly extended.

In the future, when quiescence is removed entirely, this responsibility
should be moved to the Raft tick loop.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: register lease renewals on splits/merges

The store lease renewer, which eagerly extends expiration-based leases,
did not correctly (un)register ranges on splits/merges/relocations. This
could cause the lease renewer to not eagerly extend leases for some
ranges, but was mitigated by the lease renewer only being used for the
meta and liveness ranges which rarely/never split.

This patch makes sure the lease renewer is notified on these events. It
also incidentally fixes an unlikely race where a range split immediately
following an expiration-based lease transfer could leave the RHS using
an expiration-based lease until it was transferred elsewhere.

Epic: none
Release note: None

erikgrinaker committed:
kvserver: eagerly extend expiration leases in Raft scheduler

This patch eagerly extends expiration leases in the Raft scheduler,
during Raft ticks, but only if `kv.expiration_leases_only.enabled` is
`true`. This is possible because we no longer allow ranges
with expiration leases to quiesce in this case.

The old store lease renewer is disabled in this case, but still tracks
local system ranges with expiration leases in case the setting is later
changed. The store lease renewer is still retained for use by default,
to allow backporting this change to 23.1 while keeping the existing
default behavior.

Doing this during Raft scheduling is compelling, since it already does
efficient goroutine scheduling (including liveness range prioritization)
and holds the replica mutex.

Compared to an alternative implementation that relied on a separate,
improved store lease renewer which also allowed ranges to quiesce
between extensions, this has significantly better performance. The main
reason for this is that unquiescing involves a Raft append to wake the
leader. On an idle cluster with 20.000 ranges using only expiration
leases, CPU usage was 26% vs.  30%, write IOPS was 500 vs. 900, and
network traffic was 0.8 MB/s vs.  1.0 MB/s. KV benchmark results:

```
name                    old ops/sec  new ops/sec  delta
kv0/enc=false/nodes=3    14.6k ± 3%   15.2k ± 4%  +3.72%  (p=0.003 n=8+8)
kv95/enc=false/nodes=3   30.6k ± 3%   32.2k ± 3%  +5.22%  (p=0.000 n=8+8)

name                    old p50      new p50      delta
kv0/enc=false/nodes=3     11.3 ± 3%    11.0 ± 0%  -2.76%  (p=0.006 n=8+6)
kv95/enc=false/nodes=3    4.59 ± 8%    4.55 ± 3%    ~     (p=0.315 n=8+8)

name                    old p95      new p95      delta
kv0/enc=false/nodes=3     31.5 ± 0%    30.2 ± 4%  -4.25%  (p=0.003 n=6+8)
kv95/enc=false/nodes=3    19.5 ± 7%    18.1 ± 4%  -7.28%  (p=0.000 n=8+7)

name                    old p99      new p99      delta
kv0/enc=false/nodes=3     48.7 ± 3%    47.4 ± 6%    ~     (p=0.059 n=8+8)
kv95/enc=false/nodes=3    32.8 ± 9%    30.2 ± 4%  -8.04%  (p=0.001 n=8+8)
```

Epic: none
Release note: None

erikgrinaker committed:
kvserver: conditionally extend expiration leases during requests

If `kv.expiration_leases_only.enabled´ is `true`, this patch disables
expiration lease extension during request processing, since the Raft
scheduler will eagerly extend leases in this case. This will eventually
be removed completely, but is kept for now with an eye towards a 23.1
backport.

Epic: none
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/100686)
<!-- Reviewable:end -->

erikgrinaker committed:
kvserver: deflake `TestLeaseRenewer`

Epic: none
Release note: None

erikgrinaker commented:
Need to sort out some test flakes before merging this, but won't have time until next week. See #100693 and #100689.

erikgrinaker committed:
kvserver: fix lease renewal registration of expired leases

The store lease renewer begins tracking an system range's
expiration-based lease for renewal when it's first applied. However, it
only did this if the lease was still valid when first applied. It was
possible for the node to apply a lease after it's lapsed, but then to
retain and extend the lease, and in this case it would not be registered
with the lease renewer.

Epic: none
Release note: None
